src/Surfnet/StepupRa/RaBundle/Security/Authentication/Token/SamlToken.php

```diff
@@ -39,18 +39,19 @@ class SamlToken extends AbstractToken
      */
     private $institutionConfigurationOptions;
 
-    public function __construct(
-        Loa $loa,
-        array $roles = [],
-        InstitutionConfigurationOptions $institutionConfigurationOptions = null
-    ) {
+    public function __construct(Loa $loa, array $roles = [])
```

Comment:
w.r.t. the controller: the tokenstorage [is already available][1], you modify an object by reference, therefor [setting the token][2] is not needed, etc. Various small fixes due to being forced to handle it differently.

```php
public function selectInstitutionAction(Request $request)
{
    $this->denyAccessUnlessGranted(['ROLE_SRAA']);

    $logger = $this->get('logger');
    /** @var SamlToken $token */
    $token = $this->get('security.token_storage')->getToken();
    /** @var Identity $identity */
    $identity = $token->getUser();

    $logger->notice(sprintf('Select Institution for SRAA "%s"', $identity->id));

    $command = new SelectInstitutionCommand();
    $command->institution = $identity->institution;

    $form = $this->createForm('sraa_institution_select', $command);
    $form->handleRequest($request);

    if ($form->isValid()) {
        $institutionConfigurationOptions = $this->get('ra.service.institution_configuration_options')
            ->getInstitutionConfigurationOptionsFor($command->institution);

        $token->changeInstitutionScope(
            new Institution($command->institution),
            $institutionConfigurationOptions
        );

        $flashMessage = $this->get('translator')
            ->trans('ra.sraa.changed_institution', ['%institution%' => $command->institution]);

        $this->get('session')->getFlashBag()->add('success', $flashMessage);

        $logger->notice(sprintf(
            'SRAA "%s successfully switched to institution "%s"',
            $identity->id,
            $newInstitution
        ));

        return $this->redirect($this->generateUrl('ra_vetting_search'));
    }

    $logger->notice(sprintf('Showing select institution form for SRAA "%s"', $identity->id));

    return $this->render(
        'SurfnetStepupRaRaBundle:Sraa:selectInstitution.html.twig',
        ['form' => $form->createView()]
    );
}
```

Something like that :smile:

W.r.t. the command handler, the reason I mentioned that for this controller specific is that we're doing two actions in one place as a consequence of a single command (override institution, but also fetch InstConfOptions and override those), which both lead to a change of a single object (the token). This is an indicator a service level object could be introduced to move the complexity of state manipulation from the controller ("infra") to the "application". Not necessarily a tech debt, more a clean way to deal with increased complexity and requirements for state change based on the interaction with the user. Heh, sounds fancy. In any case, was just a comment, not saying it must be done, and certainly not across the board (albeit most controllers already use some form of command handling, esp. on write actions :wink: )

[1]: https://github.com/SURFnet/Stepup-RA/pull/119/files#diff-94e8dd91808b5b0de0f08d6631f77728R35
[2]: https://github.com/SURFnet/Stepup-RA/pull/119/files#diff-94e8dd91808b5b0de0f08d6631f77728R60